plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs

```diff
@@ -0,0 +1,118 @@
+-- This is not a problem as the CEK machines handles name capture by design.
+
+module PlutusCoreInterpreter.CekMachine
+    ( EvaluationResult (..)
+    , evaluateCek
+    , runCek
+    ) where
+
+import           Language.PlutusCore
+import           Language.PlutusCore.Constant
+import           Language.PlutusCore.Evaluation.MachineException (MachineError (..), MachineException (..))
+import           Language.PlutusCore.Evaluation.Result           (EvaluationResult (..))
+import           Language.PlutusCore.View
+import           PlutusPrelude
+
+import           Data.IntMap                                     (IntMap)
+import qualified Data.IntMap                                     as IntMap
+
+type Plain f = f TyName Name ()
+
+-- | Environments used by the CEK machine.
+-- Each row is a mapping from the 'Unique' representing a variable to the 'Value' the
+-- variable stands for and the environment the value was defined in.
+newtype Environment = Environment (IntMap (Plain Value, Environment))
+
+data Frame
+    = FrameApplyFun Environment (Plain Value)    -- ^ @[V _]@
+    | FrameApplyArg Environment (Plain Term)     -- ^ @[_ N]@
+    | FrameTyInstArg (Type TyName ())            -- ^ @{_ A}@
+    | FrameUnwrap                                -- ^ @(unwrap _)@
+    | FrameWrap () (TyName ()) (Type TyName ())  -- ^ @(wrap α A _)@
+
+type Context = [Frame]
+
+-- | Extend an environment with a variable name, the value the variable stands for
+-- and the environment the value is defined in.
+extendEnvironment :: Name () -> Plain Value -> Environment -> Environment -> Environment
+extendEnvironment argName arg argEnv (Environment oldEnv) =
+    Environment $ IntMap.insert (unUnique $ nameUnique argName) (arg, argEnv) oldEnv
+
+-- | Look up a name in an environment.
+lookupName :: Name () -> Environment -> Maybe (Plain Value, Environment)
+lookupName name (Environment env) = IntMap.lookup (unUnique $ nameUnique name) env
+
+-- | The computing part of the CEK machine.
+computeCek :: Environment -> Context -> Plain Term -> EvaluationResult
+computeCek env con (TyInst _ fun ty)      = computeCek env (FrameTyInstArg ty : con) fun
+computeCek env con (Apply _ fun arg)      = computeCek env (FrameApplyArg env arg : con) fun
+computeCek env con (Wrap ann tyn ty term) = computeCek env (FrameWrap ann tyn ty : con) term
+computeCek env con (Unwrap _ term)        = computeCek env (FrameUnwrap : con) term
+computeCek env con tyAbs@TyAbs{}          = returnCek env con tyAbs
+computeCek env con lamAbs@LamAbs{}        = returnCek env con lamAbs
+computeCek env con constant@Constant{}    = returnCek env con constant
+computeCek _   _   Error{}                = EvaluationFailure
+computeCek env con var@(Var _ name)       = case lookupName name env of
+    Nothing           -> throw $ MachineException OpenTermEvaluatedMachineError var
+    Just (term, env') -> returnCek env' con term
+
+-- | The returning part of the CEK machine.
```

Comment:
I found it helpful to realise that this function always pops a frame off the stack (unless it's done).